src/Service/Image.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+namespace App\Service;
+
+use App\Exception\FilesystemException;
+use App\Exception\ImageFetchException;
+use Liip\ImagineBundle\Message\WarmupCache;
+use Symfony\Component\Filesystem\Exception\IOExceptionInterface;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Filesystem\Path;
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+final class Image implements ImageInterface
+{
+    public function __construct(
+        private readonly HttpClientInterface $client,
+        private readonly string $publicPath,
+        private readonly MessageBusInterface $messageBus,
+    ) {
+    }
+
+    /**
+     * @throws FilesystemException
+     * @throws TransportExceptionInterface
+     * @throws ImageFetchException
+     */
+    public function fetch(string $url): string
+    {
+        $filesystem = new Filesystem();
+
+        try {
+            $path = $this->generatePath(url: $url, absolute: true);
+            $filesystem->mkdir(
+                Path::normalize($path),
+                0775
+            );
+        } catch (IOExceptionInterface $exception) {
+            throw new FilesystemException('Unable to create upload folder', $exception->getCode(), $exception);
+        }
+
+        $response = $this->client->request('GET', $url);
+        if (200 !== $response->getStatusCode()) {
+            throw new ImageFetchException(sprintf('Failed to fetch %s with code %s', $url, $response->getStatusCode()), $response->getStatusCode());
+        }
+
+        $dest = $path.basename($url);
```

Comment:
We should sanitize the filename; `basename('https://placehold.co/800?text=Hello+World&font=roboto')`, say, is probably not at filename we want to use.

A hash of the URL and maybe an extension based on the mime type returned in the response is one way to do it (cf. https://symfony.com/doc/current/components/mime.html#mime-types-utilities).

Reply:
Fixed